File: Text_Editor.py
import os
import subprocess
from colorama import init, Fore,  Style, Back

class ConsoleBasedTextEditor:
    

    def __init__(self):
        self.filepath=""
        self.text=[]

 
    def open_file(self, path):

        try:
            with open(path, 'r') as file:
                self.text=file.readlines()
                self.filepath=path
                print("File is opened successfully.")
 
        except FileNotFoundError:
            print("Error reading the file.")

    
    def save_file(self):
        
        try:
            with open(self.filepath, "w") as file:
                file.writelines(self.text)

            print("\nFile saved successfully.")
            self.open_in_notepad()

        except Exception:
            print("\nError saving the file.")


    def edit_file(self, line_num, new_text):
        
            while len(self.text)<=line_num:
                self.text.append("\n")

            self.text[line_num]=self.text[line_num].strip()+" "+new_text+"\n"
            


    def delete_line(self, line_num):
        
        try:
            self.text.pop(line_num)
            print("Line",line_num+1,"is deleted successfully.")

        except IndexError:
            print("Invalid line number")


    def delete_word(self, word):
        
        found=self.find_in_file(word)

        if not found:
            return 
        
        if len(word)==1 and word.isalpha():
            positions_info=[(i+1,[pos+1 for pos, char in enumerate(line) if char==word]) for i, line in enumerate(self.text) if word in line]

            if not positions_info:
                print(word + " not found as a single character in any line.")
                return


            try:

                line_choice = int(input("Choose a line number to delete the letter from: ")) - 1
                valid_lines = [line_num - 1 for line_num, _ in positions_info]

                if line_choice not in valid_lines:
                    print("Invalid line number.")
                    return

                selected_line_positions = next(positions for line_num, positions in positions_info if line_num == line_choice + 1)
                position_choice = int(input("Choose a position to delete: ")) - 1

                if (position_choice + 1) in selected_line_positions:
                    
                    line = self.text[line_choice]
                    self.text[line_choice] = line[:position_choice] + line[position_choice + 1:]
                    print("\nDeleted " + word + " from line", line_choice + 1, "at position", position_choice + 1,".")
                
                else:
                    print("Invalid position choice.")
            
            except ValueError:
                print("Invalid input. Please enter a valid number.")
        
        else:

            for i, line in enumerate(self.text):

                if word in line:  
                    print("\nFound " + word + " in line", i + 1, ": " + line.strip())

                    if word in line.split():  

                        self.text[i] = " ".join(w for w in line.split() if w != word) + "\n"
                        print("Deleted exact match " + word + " from line", i + 1)

                    else:  

                        modified_line = line.replace(word, "")
                        self.text[i] = modified_line
                        print("Deleted part of the word " + word + " in line", i + 1)
                    return

            print("\n'" + word + " not found.")
            

    def find_in_file(self, text_to_find):

        found=False

        if len(text_to_find)==1 and text_to_find.isalpha():

                for i, line in enumerate(self.text):
                    position=[pos+1 for pos, char in enumerate(line) if char==text_to_find]
                    
                    if position:
                        print("Found ",text_to_find, "in line",i+1,"at position", position)
                        found=True
        
        elif text_to_find.strip() and len(text_to_find.split())==1:
            
            for i, line in enumerate(self.text):
                words=line.split()
                position=[pos+1 for pos, word in enumerate(words) if word==text_to_find]

                if position:
                    print("Found "+text_to_find,"in line",i+1,"at position",position)
                    found=True

                else: 
                    partial_matches = [word for word in words if text_to_find in word]
                    
                    if partial_matches:
                        print(text_to_find+" is part of the word(s)",partial_matches,"in line",i + 1)
                        found = True

        else: 

            for i, line in enumerate(self.text):
                
                if text_to_find in line:
                    print("Found "+text_to_find,"in line",i+1)
                    found=True

        if not found:
            print("Content "+text_to_find,"not found.")
        
        return found


    def replace_word(self, word, new_word):
        
        found=self.find_in_file(word)

        if not found: 
            return 

        if len(word) == 1 and word.isalpha():

            positions_info = [ (i + 1, [pos + 1 for pos, char in enumerate(line) if char == word])
                for i, line in enumerate(self.text) if word in line]

            if not positions_info:
                print(word,"not found as a single character in any line.")
                return

            try:
                line_choice = int(input("Choose a line to replace the letter: ")) - 1
                valid_lines = [line_num - 1 for line_num, _ in positions_info]

                if line_choice not in valid_lines:
                    print("Invalid line number.")
                    return

                selected_line_positions = next(positions for line_num, positions in positions_info if line_num == line_choice + 1)
                position_choice = int(input("Choose a position to replace: ")) - 1

                if (position_choice + 1) in selected_line_positions:

                    line = self.text[line_choice]
                    self.text[line_choice] = line[:position_choice] + new_word + line[position_choice + 1:]
                    print("The letter",word,"has been replaced with",new_word,"in line",line_choice + 1)

                else:
                    print("Invalid position choice.")

            except ValueError:
                print("Invalid input. Please enter a valid number.")


        else: 
            multi_line_word=[i for i, line in enumerate(self.text) if word in line]

            if len(multi_line_word)>1:

                line_num=int(input("Enter the line of the word you want to replace: "))-1

                if line_num in multi_line_word:
                    self.text[line_num]=self.text[line_num].replace(word,new_word)
                    print("The word",word,"is replaced with the word:",new_word)
                    return 

                else:
                    print("The word",word,"is not found in the line!")
                    return

            else:
                for i, line in enumerate(self.text):
                    self.text[i]=line.replace(word,new_word)

            print("The word",word,"is replaced with the word:",new_word)


    def display_text(self):
        
        for i, line in enumerate(self.text, start=1):

            print(line.strip())


    def open_in_notepad(self):
        
        if os.name == 'nt' and self.filepath:
            subprocess.Popen(['notepad.exe', self.filepath])
        else:
            print("\nNotepad is not available!")


    def start_program(self):
        
        init()
        
        print(Fore.RED+"\033[1m**Text Editor**"+Style.RESET_ALL)
        print(Fore.RED+"1. "+Style.RESET_ALL+"Open File")
        print(Fore.RED+"2. "+Style.RESET_ALL+"Save File")
        print(Fore.RED+"3. "+Style.RESET_ALL+"Edit File")
        print(Fore.RED+"4. "+Style.RESET_ALL+"Find")
        print(Fore.RED+"5. "+Style.RESET_ALL+"Replace")
        print(Fore.RED+"6. "+Style.RESET_ALL+"Delete Line")
        print(Fore.RED+"7. "+Style.RESET_ALL+"Delete Word")
        print(Fore.RED+"8. "+Style.RESET_ALL+"Display File")
        print(Fore.RED+"9. "+Style.RESET_ALL+"Exit")
        
        while True:
            
            commands= int(input(Fore.RED+'\nChoose an option: '+Style.RESET_ALL))

            if commands==1:

                path=input("\nEnter the path of the file you want to open: ").strip()
                self.open_file(path)


            elif commands==2:

                if not self.text:
                    print("\nNo content to save. Please open or edit a file first.")

                else:
                    self.filepath
                    self.save_file()
            

            elif commands==3:

                if not self.text:
                    print("\nThe file is not opened or it has no content.")

                else:
                    
                    print("\nEditing opened")

                    while(True):

                        line_num=input("Enter the line number to edit (or ':x' to exit): ").strip()
                            
                        if line_num==':x':

                            print("Edit exited.")
                            break
                        
                        if not line_num.isdigit():
                            print("Invalid input. Please enter a valid line number.")
                            continue

                        line_num=int(line_num)-1
                        new_text=input("Enter the content to add: ").strip()
                        self.edit_file(line_num, new_text)
                        print("Content added successfully.")
                            

            elif commands==4:

                if not self.text:
                    print("\nThe file is not opened or it has no content.")

                else:
                     words=input("\nEnter the content you want to find: ").strip()
                     self.find_in_file(words)
                     
                
            elif commands==5:

                if not self.text:
                    print("\nThe file is not opened or it has no content.")

                else:

                     word=input("\nEnter the word to replace: ").strip()
                     new_word=input("Enter the new word: ").strip()
                     self.replace_word(word,new_word)



            elif commands==6:

                if not self.text:
                    print("\nThe file is not opened or it has no content.")

                else:
                    line_num=int(input("\nEnter the line number to delete: "))
                    self.delete_line(line_num-1)


            elif commands==7:
                
                if not self.text:
                    print("\nThe file is not opened or it has no content.")

                else:
                     word=input("\nEnter the word to delete: ").strip()
                     self.delete_word(word)


            elif commands==8:

                if not self.text:
                    print("\nThe file is not opened or it has no content.") 
                else:
                    print("Current content:")
                    print("\n")
                    self.display_text()


            elif commands==9:
                print(Fore.RED+"\nThank you for using our text editor :)\n"+Style.RESET_ALL)
                break


            else:
                print("Invalid option. Please try again.")


e=ConsoleBasedTextEditor()
e.start_program()

File: textEditor.js
//Console-based Text Editor

//import neccessary modules (first one for working with files, second one for user input, and the third one for color)
import fs from 'fs';
import readline from 'readline';
import chalk from 'chalk';

//create an interface that allows reading user input from the terminal
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

//menu options
function menu() {
    console.log(chalk.bold.red("\n** Text Editor **"));
    console.log(chalk.red("1.") + " Open File");
    console.log(chalk.red("2.") + " Save File");
    console.log(chalk.red("3.") + " Edit File");
    console.log(chalk.red("4.") + " Find");
    console.log(chalk.red("5.") + " Replace");
    console.log(chalk.red("6.") + " Delete Line");
    console.log(chalk.red("7.") + " Delete Word");
    console.log(chalk.red("8.") + " Display File");
    console.log(chalk.red("9.") + " Exit");
    rl.question(chalk.red("\nChoose an option: "), menuOptions);
}

//global variable that will store and modify the content of our file
let fileContent = '';

//global variable that checks if the file is opened
let fileOpened = false;

//global variable that stors the path of the opened file
let openedFilePath = ''

//handling possible menu options
function menuOptions(option) {
    if (!fileOpened && (option === '2' || option === '3' || option === '4' || option === '5' || option === '6' || option === '7' || option === '8')) {
        console.log("Please open a file first.");
        return rl.question(chalk.red("\nChoose an option: "), menuOptions);
    }

    switch (option) {
        case '1':
            openFile();
            break;
        case '2':
            saveFile();
            break;
        case '3':
            editFile();
            break;
        case '4':
            findWord();
            break;
        case '5':
            replaceWord();
            break;
        case '6':
            deleteLine();
            break;
        case '7':
            deleteWord();
            break;
        case '8':
            displayFile();
            break;
        case '9':
            console.log(chalk.bold.red("Thank you for using our text editor :)\n"));
            rl.close();
            break;
        default:
            console.log("Invalid option. Please try again.");
            rl.question(chalk.red("\nChoose an option: "), menuOptions);
    }
}

//opening a file
function openFile() {
    rl.question("Enter the path of the file you want to open: ", (path) => {
        try {
            if (!fs.existsSync(path)) {
                console.error("The path is invalid or the file does not exist. Please try again.");
                rl.question(chalk.red("\nChoose an option: "), menuOptions);
                return;
            }

            fileContent = fs.readFileSync(path, 'utf8');
            openedFilePath = path;
            fileOpened = true;
            console.log("File opened successfully.");
        }
        catch (error) {
            console.error("Error reading the file. Please ensure the file path is correct.");
        }
        finally {
            rl.question(chalk.red("\nChoose an option: "), menuOptions);
        }
    });
}

//saving a file
function saveFile() {
    if (!fileOpened) {
        console.log("No file has been opened or edited yet. Please open a file first.");
        return rl.question(chalk.red("\nChoose an option: "), menuOptions);
    }

    rl.question("Enter the path to save the file: ", (path) => {
        try {
            if (!fs.existsSync(path)) {
                console.error("The path is invalid or the file does not exist.");
                return rl.question(chalk.red("\nChoose an option: "), menuOptions);
            }

            fs.writeFileSync(path, fileContent, 'utf8');
            console.log("File saved successfully.");
        }
        catch (error) {
            console.error("Error saving the file. Please ensure the file path is correct.");
        }
        finally {
            rl.question(chalk.red("\nChoose an option: "), menuOptions);
        }
    });
}

//editing in the editor
function editFile() {
    console.log("Editing opened.");
    const lines = fileContent.split("\n");

    function editingLoop() {
        rl.question("Enter the line number to edit (':x' to exit editing): ", (lineNumber) => {
            try {
                if (lineNumber === ":x") {
                    console.log("Editing exited.");
                    return rl.question(chalk.red("\nChoose an option: "), menuOptions);
                }

                lineNumber = parseInt(lineNumber);
                if (isNaN(lineNumber) || lineNumber < 1) {
                    console.log("Invalid line number.");
                    editingLoop();
                    return;
                }

                rl.question("Enter the content to add: ", (content) => {
                    while (lines.length < lineNumber) {
                        lines.push("");
                    }
                    lines[lineNumber - 1] += (lines[lineNumber - 1] ? " " : "") + content;
                    fileContent = lines.join("\n");
                    console.log("Content added successfully.");
                    editingLoop();
                });
            }
            catch (error) {
                console.error("An error occurred while editing. Please try again.");
                editingLoop();
            }
        });
    }
    editingLoop();
}

//finding a specific word
function findWord() {
    try {
        if (!fileContent) {
            console.log("No content in the file.");
        }
        else {
            rl.question("Enter the word to find: ", (word) => {
                const lines = fileContent.split("\n");
                let found = false;
                lines.forEach((line, index) => {
                    let position = line.indexOf(word);
                    while (position !== -1) {
                        console.log(`Found '${word}' in line ${index + 1} at position ${position + 1}`);
                        position = line.indexOf(word, position + 1);
                        found = true;
                    }
                });
                if (!found) {
                    console.log(`Word '${word}' not found.`);
                }
                rl.question(chalk.red("\nChoose an option: "), menuOptions);
            });
        }
    }
    catch (error) {
        console.error("An error occurred while searching for the word.");
        rl.question(chalk.red("\nChoose an option: "), menuOptions);
    }
}

//replacing a word
function replaceWord() {
    if (!fileContent) {
        console.log("No content in the file.");
        rl.question(chalk.red("\nChoose an option: "), menuOptions);
        return;
    }

    rl.question("Enter the word to replace: ", (oldWord) => {
        const lines = fileContent.split("\n");
        const occurrences = [];

        lines.forEach((line, index) => {
            let startPosition = 0;
            while ((startPosition = line.indexOf(oldWord, startPosition)) !== -1) {
                occurrences.push({ line: index + 1, position: startPosition + 1 });
                console.log(`Found '${oldWord}' in line ${index + 1} at position ${startPosition + 1}`);
                startPosition += oldWord.length;
            }
        });

        if (occurrences.length === 0) {
            console.log(`Word '${oldWord}' not found in the file.`);
            rl.question(chalk.red("\nChoose an option: "), menuOptions);
            return;
        }

        rl.question("Enter the line number where you want to replace the word: ", (lineNumberInput) => {
            const lineNumber = parseInt(lineNumberInput);
            if (isNaN(lineNumber) || lineNumber < 1 || lineNumber > lines.length) {
                console.log("Invalid line number. Returning to menu.");
                rl.question(chalk.red("\nChoose an option: "), menuOptions);
                return;
            }

            rl.question("Enter the position of the word to replace: ", (positionInput) => {
                const position = parseInt(positionInput);
                if (isNaN(position) || position < 1) {
                    console.log("Invalid position. Returning to menu.");
                    rl.question(chalk.red("\nChoose an option: "), menuOptions);
                    return;
                }

                const line = lines[lineNumber - 1];
                const occurrencesInLine = [];
                let startPosition = 0;

                while ((startPosition = line.indexOf(oldWord, startPosition)) !== -1) {
                    occurrencesInLine.push(startPosition + 1);
                    startPosition += oldWord.length;
                }

                if (!occurrencesInLine.includes(position)) {
                    console.log(`The word '${oldWord}' is not found at position ${position} in line ${lineNumber}. Returning to menu.`);
                    rl.question(chalk.red("\nChoose an option: "), menuOptions);
                    return;
                }

                rl.question("Enter the new word to replace it with: ", (newWord) => {
                    const beforeWord = line.slice(0, position - 1);
                    const afterWord = line.slice(position - 1 + oldWord.length);
                    lines[lineNumber - 1] = beforeWord + newWord + afterWord;
                    fileContent = lines.join("\n");
                    console.log(`Successfully replaced '${oldWord}' with '${newWord}' on line ${lineNumber}, position ${position}.`);
                    rl.question(chalk.red("\nChoose an option: "), menuOptions);
                });
            });
        });
    });
}

//deleting a specific line
function deleteLine() {
    try {
        if (!fileContent) {
            console.log("No content in the file.");
        }
        else {
            const lines = fileContent.split("\n");
            rl.question("Enter the line number to delete: ", (lineNumber) => {
                lineNumber = parseInt(lineNumber);
                if (isNaN(lineNumber) || lineNumber < 1 || lineNumber > lines.length) {
                    console.log("Invalid line number.");
                } else {
                    lines.splice(lineNumber - 1, 1);
                    fileContent = lines.join("\n");
                    console.log(`Line ${lineNumber} deleted successfully.`);
                }
                rl.question(chalk.red("\nChoose an option: "), menuOptions);
            });
        }
    }
    catch (error) {
        console.error("An error occurred while deleting the line.");
        rl.question(chalk.red("\nChoose an option: "), menuOptions);
    }
}

//deleting a specific word
function deleteWord() {
    if (!fileContent) {
        console.log("No content in the file.");
        rl.question(chalk.red("\nChoose an option: "), menuOptions);
        return;
    }

    const lines = fileContent.split("\n");

    rl.question("Enter the word to delete: ", (word) => {
        let found = false;
        const occurrences = [];

        lines.forEach((line, lineIndex) => {
            let startPosition = 0;
            while ((startPosition = line.indexOf(word, startPosition)) !== -1) {
                occurrences.push({
                    lineNumber: lineIndex + 1,
                    position: startPosition + 1,
                });
                console.log(`Found '${word}' in line ${lineIndex + 1} at position ${startPosition + 1}`);
                startPosition += word.length;
                found = true;
            }
        });

        if (!found) {
            console.log(`Word '${word}' not found in the file.`);
            rl.question(chalk.red("\nChoose an option: "), menuOptions);
            return;
        }

        rl.question("Enter the line number where you want to delete the word: ", (lineNumberInput) => {
            const lineNumber = parseInt(lineNumberInput);

            if (isNaN(lineNumber) || lineNumber < 1 || lineNumber > lines.length) {
                console.log("Invalid line number. Returning to menu.");
                rl.question(chalk.red("\nChoose an option: "), menuOptions);
                return;
            }

            rl.question("Enter the position of the word to delete: ", (positionInput) => {
                const position = parseInt(positionInput);

                const occurrence = occurrences.find(
                    (occ) => occ.lineNumber === lineNumber && occ.position === position
                );

                if (!occurrence) {
                    console.log("Invalid position. Returning to menu.");
                    rl.question(chalk.red("\nChoose an option: "), menuOptions);
                    return;
                }

                const line = lines[lineNumber - 1];
                const beforeWord = line.slice(0, position - 1);
                const afterWord = line.slice(position - 1 + word.length);
                lines[lineNumber - 1] = beforeWord + afterWord;

                fileContent = lines.join("\n");
                console.log(`Successfully deleted '${word}' from line ${lineNumber}, position ${position}.`);
                rl.question(chalk.red("\nChoose an option: "), menuOptions);
            });
        });
    });
}

//displaying all the contents of the file
function displayFile() {
    try {
        if (!fileContent) {
            console.log("No content in the file.");
        } else {
            console.log("Current content: \n" + fileContent);
        }
    } catch (error) {
        console.error("An error occurred while displaying the file.");
    } finally {
        rl.question(chalk.red("\nChoose an option: "), menuOptions);
    }
}

menu();
